src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -796,8 +796,11 @@ int dhcp_server_handle_message(sd_dhcp_server *server, DHCPMessage *message,
                         r = sd_event_now(server->event,
                                          clock_boottime_or_monotonic(),
                                          &time_now);
-                        if (r < 0)
+                        if (r < 0) {
+                                dhcp_lease_free(lease);
```

Comment:
This is only valid if `lease` wasn't assigned to `existing_lease` above. Otherwise you're freeing a pointer that is still stored in the hash map. The correct fix is to repeat what's used below:

```c
if (r < 0) {
    if (!existing_lease)
        dhcp_lease_free(lease);
    return r;
}
```